File: D3/kMC/Ex01/Ex01.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* ------- VARIABLE DECLARATION...
   ------------------------------- */

// ::: PHYSICAL PARAMETER
double kbev = 8.61733e-5;
double kb = 1.38e-23;
double na = 6.02214e23;
double pi = 4.0 * atan(1.0);
double u = 1.66e-27;
double ang = 1e-10;

// ::: USER VARIABLE
int save_site;

// ::: STRUCTURE
struct event_type {

  int nevent, nbond, nchem_react;

  int *ptr_i_state, // array contains Initial state
      *ptr_f_state; // array contains Final state

  double *ptr_f0,    // array contains prefactor rate
      *ptr_ebarrier, // array contains activation energies
      *ptr_de,       // array contains difference energy between two miniums
      **ptr_ebond;   // 2D array contains bond energies
};

struct kmc_type {

  int bavard, conv;         // Verbose & convergence
  int period[3], nsites[3]; // Bound periodic condition & number of site

  char algorithm[50],  // algorithm name: BKL or Gillepsie
      input_file[50],  // input_file name
      input_event[50], // event file name
      libname[50],     // shared lib name with the path
      init_mod[50];    // Initialization mode: species, random, none

  int tot_sites,  // Total number of sites
      max_step,   // max step criterium
      sys_dim,    // system dimension, 1, 2 or 3 D
      freq_write, // write stat frequency
      nprop,      // number of properties compute in analyse routine
      node_state, // number of state of one site
      nspec,      // number of different sepcies
      npressure,  // number of partial pressure
      step;       // actual step

  double sum_rate, // Total rate of system
      rand_rate,   // random rate
      time,        // actual time
      rand_time,   // random time
      max_time,    // max time criterium
      temp,        // temperature value (Kelvin)
      kt,          // Energy of temperature
      per100,      // Percentage of 0 state in initial system
      f0,          // prefactor rate value
      scale;       // scale value

  int *ptr_site, *ptr_nneig, *ptr_nevt, *ptr_neig, *ptr_event_site, *ptr_spec;
  double *ptr_rate, *ptr_prop, *ptr_event_rate, *ptr_pressure, *ptr_masse;

  struct event_type event;
};

extern int *__hidden_table_MOD_h_i_state, *__hidden_table_MOD_h_f_state;
extern double *__hidden_table_MOD_h_ebarrier, *__hidden_table_MOD_h_de,
    **__hidden_table_MOD_h_ebond, *__hidden_table_MOD_h_f0;

/* ------- FUNCTION DECLARATION...
   ------------------------------- */

void builder_event_type(struct event_type *, int);

/* ------- FUNCTION DEFINITION...
   ------------------------------ */

char **parsing(char *string, int len) {
  int i = 0;
  char **pch;
  pch = malloc(len * sizeof(char));
  pch[i] = strtok(string, "\t ");
  while (pch[i] != NULL) {
    // printf( "sparse %d %s \n",i, pch[i] );
    i++;
    pch[i] = strtok(NULL, "\t ");
  }
  return pch;
}
// ...........................................................

double random_number(double min, double max) {
  return min + (double)rand() / ((double)(RAND_MAX + 1.0) * (max - min));
}
// ......................................................................
// .................................................................................
// ......................................................................

void builder_event_type(struct event_type *event, int nevent) {
  int int_size = sizeof(int);
  int double_size = sizeof(double);

  printf(" Enter in EVENT_type constructor... %d\n", nevent);
  event->nevent = nevent;

  __hidden_table_MOD_h_i_state = malloc(nevent * int_size);
  __hidden_table_MOD_h_f_state = malloc(nevent * int_size);
  __hidden_table_MOD_h_f0 = malloc(nevent * double_size);
  __hidden_table_MOD_h_ebarrier = malloc(nevent * double_size);
  __hidden_table_MOD_h_de = malloc(nevent * double_size);

  if (event->nbond != 0) {
    __hidden_table_MOD_h_ebond = malloc(event->nbond * sizeof(double *));
    for (int i = 0; i < event->nbond; i++)
      __hidden_table_MOD_h_ebond[i] = malloc(event->nbond * sizeof(double **));
  }

  event->ptr_i_state = __hidden_table_MOD_h_i_state;
  event->ptr_f_state = __hidden_table_MOD_h_f_state;
  event->ptr_f0 = __hidden_table_MOD_h_f0;
  event->ptr_ebarrier = __hidden_table_MOD_h_ebarrier;
  event->ptr_de = __hidden_table_MOD_h_de;
  if (event->nbond != 0)
    event->ptr_ebond = __hidden_table_MOD_h_ebond;

  printf(" EVENT_type constructor DONE \n");
}
// ...........................................................
//.................................................................................

void read_event(struct kmc_type *struc) {

  char string[100], **word;
  char *b = string;
  int i, id, nevent, nbond, ibd, jbd, npress;
  size_t bufsize = 100;
  size_t nword;

  FILE *fp;

  printf(" === PRINT STRUCTURE KMC ===\n");
  printf(" Periodicity %d %d %d \n", struc->period[0], struc->period[1],
         struc->period[2]);
  printf(" nsites %d %d %d\n", struc->nsites[0], struc->nsites[1],
         struc->nsites[2]);
  printf(" INPUT EVENT %s \n", struc->input_event);
  printf(" ALGORITHM   %s \n", struc->algorithm);
  printf(" INPUT FILE  %s \n", struc->input_file);
  printf(" LIBNAME     %s \n", struc->libname);

  // ::: INIT SAVE_SITE
  save_site = 0;

  // ::: READ EVENT_FILE :::

  fp = fopen(struc->input_event, "r");
  if (fp == NULL)
    perror(" PB open input_event...\n");

  nword = getline(&b, &bufsize, fp);

  while (strcmp(&string[0], "#") >= 0) {

    // SPARSE the line....
    word = parsing(string, bufsize);

    /* ==== read the event ==== */

    if (!strcmp(word[0], "Number_of_event")) {
      nevent = atoi(word[1]);
      printf(" nevent %d\n ", nevent);
      builder_event_type(&struc->event, nevent);

      for (id = 0; id < struc->event.nevent; id++) {
        nword = getline(&b, &bufsize, fp);
        word = parsing(string, bufsize);

        struc->event.ptr_i_state[id] = atoi(word[1]);
        struc->event.ptr_f_state[id] = atoi(word[2]);
        struc->event.ptr_f0[id] = atof(word[3]);
        struc->event.ptr_ebarrier[id] = atof(word[4]);
        struc->event.ptr_de[id] = atof(word[5]);
        printf(" read %d event %d %d %e %f %f \n", id,
               struc->event.ptr_i_state[id], struc->event.ptr_f_state[id],
               struc->event.ptr_f0[id], struc->event.ptr_ebarrier[id],
               struc->event.ptr_de[id]);
      }
    } // -------

    /* ==== READ PARTIAL PRESSURE ==== */

    if (!strcmp(word[0], "partial_pressure")) {

      npress = atoi(word[1]);
      if (npress != struc->npressure) {
        printf("WARNING: Number of Partial_pressure %d is different to input "
               "pressure %d\n",
               npress, struc->npressure);
        exit(1);
      }

      for (i = 0; i < npress; i++) {
        nword = getline(&b, &bufsize, fp);
        word = parsing(string, bufsize);
        id = atoi(word[0]);
        struc->ptr_pressure[id] = atof(word[1]);
        struc->ptr_masse[id] = atof(word[2]);
        printf(" PRESSURE: %d %e %f \n", id, struc->ptr_pressure[id],
               struc->ptr_masse[id]);
      }

    } // -------

    /* ==== read the energy bond ==== */

    if (!strcmp(word[0], "Energy_Bond")) {

      nbond = atoi(word[1]);

      for (id = 0; id < nbond; id++) {

        nword = getline(&b, &bufsize, fp);
        word = parsing(string, bufsize);

        ibd = atoi(word[1]);
        jbd = atoi(word[2]);
        struc->event.ptr_ebond[ibd][jbd] = atof(word[3]);
        printf(" BOND: %d %d %f --\n", ibd, jbd,
               struc->event.ptr_ebond[ibd][jbd]);
      }
    }

    // Read new line...
    nword = getline(&b, &bufsize, fp);

  } // --- END WHILE ----

  fclose(fp);

  srand(time(NULL));
}
// .................................................................................................

void event_rate_calc(struct kmc_type *obj) {}
// .................................................................................................

void choose_event(struct kmc_type *struc, int *isite, int *ievent) {

  int i, j0, jn, nevt;

  *isite = 0;
  *ievent = 0;

  double rdn = random_number(0.0, 1.0);
  double rrdn = rdn * struc->sum_rate;
  double *evt_rate = struc->ptr_event_rate;
  struc->rand_rate = rrdn;

  double rsum = 0.0;
  for (i = 0; i < struc->tot_sites; i++) {

    j0 = struc->ptr_nneig[i] - 1;
    nevt = struc->ptr_nevt[i];

    for (jn = 1; jn <= nevt; jn++) {

      *ievent = jn;
      rsum += evt_rate[j0 + jn];

      if (rsum > rrdn)
        break;
    }

    *isite = i;
    if (rsum > rrdn)
      break;
  }
  if (rsum <= rrdn)
    printf(" PB choose_event... %f %f %f %f\n", rdn, rrdn, rsum,
           struc->sum_rate);

  save_site = *isite;
}
// .................................................................................................

void event_applied(struct kmc_type *obj, int *is, int *jn) {

  int jevt, j0;

  j0 = obj->ptr_nneig[*is] - 1;
  jevt = obj->ptr_event_site[j0 + *jn];

  if (obj->ptr_site[*is] != obj->event.ptr_i_state[jevt])
    printf(" %d %d Problem site state (%d) not correspond to initial state "
           "(%d) event %d | %d \n",
           *is, *jn, obj->ptr_site[*is], obj->event.ptr_i_state[jevt], jevt,
           obj->ptr_nevt[*is]);

  obj->ptr_site[*is] = obj->event.ptr_f_state[jevt];
}
// .................................................................................................

void analyse(struct kmc_type *obj) {}
// .................................................................................................

File: D3/kMC/Ex01/Ex01.f90
  module variables
    use iso_c_binding
    implicit none
    !
    integer( c_int ), dimension(:), pointer :: site, nneig, neig, nevt, spec, event_site
    !
    real( c_double ), dimension(:), pointer :: rate, prop, event_rate, pressure, masse
    !
    ! ::: EVENT
    integer( c_int ), dimension(:), pointer :: init_state, final_state
    real( c_double ), dimension(:), pointer :: ebarrier, de, f0
    real( c_double ), dimension(:,:), pointer :: ebond
    !
    integer( c_int ) :: save_site
    !
  end module variables

! .....	.........................................................................................

    subroutine read_event( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none

      type( KMC_type ), intent( inout ) :: obj
      character (len=500,kind=c_char)   :: string
      integer( c_int )                  :: u0, i, id, ios, nevent, npress
      logical                           :: EOF
      !
      character (len=1, kind=c_char)  :: delims
      CHARACTER (len=100, kind=c_char), dimension (50) :: args
      integer( c_int ) :: nargs
      !
      !  ::: Init save_site
      !
      save_site = 0
      !
      !  ::: Lecture of state and rate of each node
      !
      open( newunit=u0, file=trim(obj% input_event), iostat=ios )
      if ( ios /= 0 ) call error( "input_event does't open!!" )
      !
      EOF = .false.
      do while ( .not.EOF )
         !
         call read_line( u0, string, EOF )
         call parse( trim(string), delims, args, nargs )
         !write (*,*) "Lu : ", nargs, (i,trim(args(i)), i=1,nargs)
         !
         ! ::: Lecture Of Event
         !
         if ( args(1) == "Number_of_event" ) then
            read( args(2), '(i5)' ) nevent
            call builder_event_type( obj% event, nevent )
            !exit
            !
            call link_int1_ptr( obj% event% ptr_i_state,   init_state,  obj% event% nevent)
            call link_int1_ptr( obj% event% ptr_f_state,   final_state, obj% event% nevent)
            call link_real1_ptr( obj% event% ptr_f0,       f0,          obj% event% nevent )
            call link_real1_ptr( obj% event% ptr_ebarrier, ebarrier,    obj% event% nevent)
            call link_real1_ptr( obj% event% ptr_de,       de,          obj% event% nevent)
            !
            do i = 1,obj% event% nevent
               !
               read (u0,*) id, init_state(id), final_state(id), f0(id), ebarrier(id), de(id)
               write (*,*) id, init_state(id), final_state(id), f0(id), ebarrier(id), de(id)
               !
            enddo
            !
         endif
         !
         ! ::: Lecture Of Partial Pressure Parameters
         !
         if ( args(1) == "partial_pressure" ) then
            !
            read( args(2), '(i5)' ) npress
            if ( npress /= obj% npressure ) call warning( "WARNING npressure /= partial_pressure" )
            !
            call link_real1_ptr( obj% ptr_pressure, pressure, obj% npressure)
            call link_real1_ptr( obj% ptr_masse, masse, obj% npressure)
            !
            do i = 1,npress
               read( u0, * ) id, pressure( id ), masse( id ) 
               print*, "Pressure", id, pressure( id ), masse( id )
            enddo
            !
         endif
         !
      enddo
      !
      !
      close( u0 )
      !stop " read_event..." 
      !
    end subroutine read_event
! .................................................................................................

    subroutine event_rate_calc( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none

      type( KMC_type ), intent( inout ) :: obj
      integer :: i
      !
      call link_int1_ptr( obj% ptr_site,             site,        obj% tot_sites )
      call link_int1_ptr( obj% ptr_nneig,            nneig,       obj% tot_sites )
      call link_int1_ptr( obj% ptr_nevt,             nevt,        obj% tot_sites )
      call link_int1_ptr( obj% event% ptr_i_state,   init_state,  obj% event% nevent )
      call link_int1_ptr( obj% event% ptr_f_state,   final_state, obj% event% nevent )
      call link_real1_ptr( obj% event% ptr_f0,       f0,          obj% event% nevent )
      call link_real1_ptr( obj% event% ptr_ebarrier, ebarrier,    obj% event% nevent )
      call link_real1_ptr( obj% event% ptr_de,       de,          obj% event% nevent )
      call link_real1_ptr( obj% ptr_rate,            rate,        obj% tot_sites )
      !
      call link_int1_ptr( obj% ptr_neig,           neig,        nvois*obj% tot_sites )
      call link_int1_ptr( obj% ptr_event_site,     event_site,  nvois*obj% tot_sites )
      call link_real1_ptr( obj% ptr_event_rate,    event_rate,  nvois*obj% tot_sites )

      call link_real1_ptr( obj% ptr_masse, masse, obj%npressure )

      !
      ! ::: We have 1 event by site : 
      !    Adsorption or Desorption
      !
      do i = 1,obj% tot_sites

      enddo
      !
      !
    end subroutine event_rate_calc
! .................................................................................................

    subroutine choose_event( struc, isite, ievent ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use random
      use variables
      implicit none
      type( KMC_type ), intent( inout ) :: struc
      integer( c_int ), intent( inout ) :: isite, ievent
      !
      integer( c_int ) :: i, jn, j0
      real( c_double ) :: rsum, rdn, rrdn
      !
      call link_int1_ptr( struc% ptr_nevt,       nevt,       struc% tot_sites )
      call link_int1_ptr( struc% ptr_nneig,     nneig,       struc% tot_sites )
      call link_real1_ptr( struc% ptr_event_rate, event_rate, nvois*struc% tot_sites )
      call link_int1_ptr( struc% ptr_event_site,   event_site, nvois*struc% tot_sites )
      !
      isite = 0 ; ievent = 0
      call random_number( rdn )
      rrdn = rdn*struc% sum_rate
      struc% rand_rate = rrdn
      !      
      rsum = 0.0
      do i = 1,struc% tot_sites
         !
         j0 = nneig( i )
         do jn = 1,nevt( i )
         !  !
            ievent = jn
            rsum = rsum + event_rate( j0 + jn )
            if ( rsum > rrdn ) exit
         !  !
         enddo
         !
         isite = i
         if ( rsum > rrdn ) exit
      enddo
      if ( rsum <= rrdn ) write (*,*) " PB choose event...", rsum,struc% sum_rate
      !
      save_site = isite
      !
    end subroutine choose_event
! .................................................................................................

    subroutine event_applied( struc, is, jn ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use errors
      use variables
      implicit none
      type( KMC_type )               :: struc
      integer( c_int ), intent( in ) :: is,   &  ! Site selected
                                        jn       ! "ievent" of site "is" selected 
      integer( c_int )               :: jevt, j0

      call link_int1_ptr( struc% ptr_site,             site,             struc% tot_sites )
      call link_int1_ptr( struc% event% ptr_i_state,   init_state,       struc% tot_sites )
      call link_int1_ptr( struc% event% ptr_f_state,   final_state,      struc% tot_sites )
      call link_int1_ptr( struc% ptr_event_site,       event_site,       nvois*struc% tot_sites )
      call link_int1_ptr( struc% ptr_nneig,            nneig,            struc% tot_sites )
      !
      j0 = nneig( is ) 
      jevt = event_site( j0 + jn )
      !
      if ( site( is ) /= init_state( jevt ) ) then
         call warning( " Problem site state not correspond to initial state event ")
      endif
      !
      site( is ) = final_state( jevt )
      !
    end subroutine event_applied
! .................................................................................................

    subroutine analyse( obj ) bind( C )
      use iso_c_binding
      use derived_types
      use sub_new_types
      use variables
      implicit none
      !
      type( KMC_type ), intent( inout ) :: obj
      !
      !
    end subroutine analyse
! .................................................................................................
















